code/game/machinery/kitchen/microwave.dm

```diff
@@ -153,237 +161,250 @@
+	var/highest_count = 0
 	for(var/rtype in all_recipes)
 		var/decl/recipe/recipe = all_recipes[rtype]
-		if(istype(recipe) && recipe.check_reagents(reagents) && recipe.check_items(src) && recipe.check_fruit(src))
-			possible_recipes += recipe
-
-	//okay, let's select the most complicated recipe
-	if(length(possible_recipes))
-		var/highest_count = 0
-		for(var/decl/recipe/recipe in possible_recipes)
-			var/count = length(recipe.items) + length(recipe.reagents) + length(recipe.fruit)
-			if(count >= highest_count)
-				highest_count = count
-				. = recipe
+		if(!istype(recipe) || !recipe.check_reagents(reagents) || !recipe.check_items(src) || !recipe.check_fruit(src))
+			continue
+		//okay, let's select the most complicated recipe
+		if(recipe.complexity >= highest_count)
+			highest_count = recipe.complexity
+			. = recipe
 
 /obj/machinery/microwave/proc/cook()
+	cook_break = FALSE
+	cook_dirty = FALSE
+
 	if(stat & (NOPOWER|BROKEN))
 		return
-	start()
-	if (reagents.total_volume==0 && !contents.len) //dry run
-		if (!wzhzhzh(10))
-			abort()
-			return
-		stop()
+
+	if (!reagents.total_volume && !length(get_contained_external_atoms())) //dry run
+		start()
 		return
 
-	var/decl/recipe/recipe = select_recipe()
-	var/obj/cooked
+	if (reagents.total_volume && prob(50)) // 50% chance a liquid recipe gets messy
+		dirty += Ceiling(reagents.total_volume / 10)
+
+	var/decl/recipe/recipe = src.select_recipe()
 	if (!recipe)
-		dirty += 1
-		if (prob(max(10,dirty*5)))
-			if (!wzhzhzh(4))
-				abort()
-				return
-			muck_start()
-			wzhzhzh(4)
-			muck_finish()
-			cooked = fail()
-			cooked.dropInto(loc)
-			return
+		failed = TRUE
+		cook_time = update_cook_time()
+		dirty += 5
+		if (prob(max(10, dirty*5)))
+			// It's dirty enough to mess up the microwave
+			cook_dirty = TRUE
 		else if (has_extra_item())
-			if (!wzhzhzh(4))
-				abort()
-				return
-			broke()
-			cooked = fail()
-			cooked.dropInto(loc)
-			return
-		else
-			if (!wzhzhzh(10))
-				abort()
-				return
-			stop()
-			cooked = fail()
-			cooked.dropInto(loc)
-			return
+			// Something's in the microwave that shouldn't be! Time to break!
+			cook_break = TRUE
 	else
-		var/halftime = round(recipe.time/10/2)
-		if (!wzhzhzh(halftime))
-			abort()
-			return
-		if (!wzhzhzh(halftime))
-			abort()
-			cooked = fail()
-			cooked.dropInto(loc)
-			return
-		cooked = recipe.make_food(src)
-		LAZYCLEARLIST(ingredients)
+		failed = FALSE
+		cook_time = update_cook_time(round(recipe.time * 2))
+
+	start()
+
+/obj/machinery/microwave/proc/update_cook_time(var/ct = 200)
+	return (ct / cooking_power)
+
+/obj/machinery/microwave/proc/finish_cooking()
+	var/decl/recipe/recipe = src.select_recipe()
+	if(!recipe)
+		return
+	var/result = recipe.result
+	var/list/cooked_items = list()
+	while(recipe)
+		cooked_items += recipe.make_food(src)
+		recipe = src.select_recipe()
+		if (!recipe || (recipe.result != result))
+			break
```

Comment:
Don't really follow: is the point that you continue applying various recipes as long as the result is the same? Why that particular condition (seems arbitrary)?

Reply:
Because it makes sense to double the recipe to make twice the amount, but trying to cook eggs and bake bread in the same container doesn't make sense.

Reply:
OK.